server/s3_head_object_action.cc

```diff
@@ -74,6 +96,10 @@ void S3HeadObjectAction::fetch_object_info_failed() {
     // There is no object list index, hence object doesn't exist
     s3_log(S3_LOG_DEBUG, request_id, "Object not found\n");
     set_s3_error("NoSuchKey");
+  } else if (!object_metadata || request->has_query_param_key("versionId")) {
```

Comment:
I guess this new check might interfere with the existing logic. E.g. with no versionId specified in the request if we reach here, the !object_metadata will be True and we will send InvalidArgument response. Please check this.

Reply:
1) !object_metadata will be true only when version_id is specified and it is empty.
2) if key is invalid it will catch in the first if condition
3)also for invalid or empty version id the error code is same
Please see below o/p
-------------------------
1)Invalid version id 
 aws s3api head-object --bucket testseagatebucket123 --key putbucketversionlog --version-id iMTg0NDY3NDI0MzM5NzAyNjc4MjUi --endpoint-url http://s3.seagate.com:80 --no-verify-ssl
An error occurred (400) when calling the HeadObject operation: Bad Request
2)empty version id
[root@ssc-vm-g2-rhev4-0907 ~]# aws s3api head-object --bucket testseagatebucket123 --key putbucketversionlog --version-id ""  --endpoint-url http://s3.seagate.com:80 --no-verify-ssl

An error occurred (400) when calling the HeadObject operation: Bad Request

3)key is empty
[root@ssc-vm-g2-rhev4-0907 ~]# aws s3api head-object --bucket testseagatebucket123 --key ""  --endpoint-url http://s3.seagate.com:80 --no-verify-ssl

Parameter validation failed:
Invalid length for parameter Key, value: 0, valid min length: 1

4)invalid key
[root@ssc-vm-g2-rhev4-0907 ~]# aws s3api head-object --bucket testseagatebucket123 --key fgdfg  --endpoint-url http://s3.seagate.com:80 --no-verify-ssl

An error occurred (404) when calling the HeadObject operation: Not Found